implement ImmutableValidator class

Marenz opened the pull request:
refs #3835

Marenz committed:
Validate immutable variables

Marenz commented:
This is just a rough draft. So far it compiles, but no tests have been done yet.
This PR is to review the general design of the class and to get early feedback.

ekpyron commented:
First response: it will need to handle recursive and mutually recursive function calls.
To that end strict ownership with a vector of ``unique_ptr``s won't work anymore. ``shared_ptr``s are bad either - they will likely become circular and never free-able. A combination of ``shared_ptr``s and ``weak_ptr``s can do this safely, but it's a hassle - so I'd suggest to just delegate ownership to some container like in ``CFG::NodeContainer``, I think that's easiest. Then you can just use plain pointers in the call graph and just keep a map of already visited function definitions to call graph nodes and reuse them during a call (you have to add the entry to the map *before* visiting the function definition though).

Also you need to consider modifiers as well.

Marenz commented:
Alright, this is a pretty good state now. Any more test cases you can think of I should add? General feedback?

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+#include <libsolidity/analysis/FunctionCallGraph.h>
+//#include <libsolutil/CommonData.h>
```
Needed?

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+#include <libsolidity/analysis/FunctionCallGraph.h>
```
License header

chriseth commented:
Can you add some tests with internal function pointers?

chriseth commented:
(I guess they don't matter much because we said we work on the references, but still)

chriseth commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+	if (!_callable->virtualSemantics())
+		return _callable;
+
+	if (auto originFuncDef = dynamic_cast<FunctionDefinition const*>(_callable))
```
This kind of duplicates code in `CompilerContext::resolveVirtualFunction` - we should extract it to a common utility class

leonardoalt replied:
Yea I agree with that. I think something similar is done also in SMTEncoder

leonardoalt commented on libsolidity/CMakeLists.txt:
```diff
@@ -14,6 +14,8 @@ set(sources
 	analysis/DocStringAnalyser.cpp
 	analysis/DocStringAnalyser.h
+	analysis/FunctionCallGraph.h
+	analysis/FunctionCallGraph.cpp
```
Please swap the lines

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+namespace solidity::frontend
+{
+
+// Traverses the constructors and state variable assignments to builds a
```
```suggestion
// Traverses the constructors and state variable assignments to build a
```

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+{
+
+// Traverses the constructors and state variable assignments to builds a
+// function call graph
```
```suggestion
// function call graph.
```

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+
+// Traverses the constructors and state variable assignments to builds a
+// function call graph
+class FunctionCallGraphBuilder : private ASTConstVisitor
```
```suggestion
class FunctionCallGraphBuilder: private ASTConstVisitor
```
not sure but I think this is the style?

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+		m_currentContract(_contractDefinition) { solAssert(_contractDefinition, "contract must be non-null."); }
+
+	void analyze();
+	CallSet const* getGraph(Declaration const* _declaration) const;
```
```suggestion
	CallSet const* graph(Declaration const* _declaration) const;
```

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+
+	CallableDeclaration const* findFinalOverride(CallableDeclaration const* _functionDefinition);
+
+	ContractDefinition const* m_currentContract;
```
```suggestion
	ContractDefinition const* m_currentContract = nullptr;
```

leonardoalt replied:
Or, does it need to be a pointer? Could it be a reference instead?

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+	Declaration const* m_currentNode = nullptr;
+	CallSet m_currentMapping;
+
+	std::map<Declaration const*, CallSet, CompareDef> m_mapping;
```
I think `m_mapping` sounds a bit vague. Could it be a bit more precise?

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.h:
```diff
@@ -0,0 +1,73 @@
+		{
+			return _lhs->id() < _rhs->id();
+		}
+		bool operator()(Declaration const* _lhs, int64_t _rhs) const
```
Why are this and the next one needed?

Marenz replied:
they allow searching the set using only the id, e.g `mySet.find(34)`

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+		{
+			m_currentNode = stateVar;
+			stateVar->value()->accept(*this);
+			m_mapping.insert(std::pair(stateVar, std::move(m_currentMapping)));
```
```suggestion
			m_mapping.emplace(stateVar, std::move(m_currentMapping));
```
not too important, but...

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+		}
+}
+
+
```
Two blanks?

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const callableDef = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
```
```suggestion
	if (auto const* callableDef = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
```

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+{
+	if (auto const callableDef = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+		if (m_currentMapping.insert(callableDef).second)
+			findFinalOverride(callableDef)->accept(*this);
```
Can `findFinalOverride` return `nullptr`?

Marenz replied:
only for declarations that don't exist in the contract which in this case is not possible

leonardoalt commented on libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -0,0 +1,109 @@
+		funcDef->body().accept(*this);
+	}
+	else if (ModifierDefinition const* modDef = dynamic_cast<decltype(modDef)>(&_callableDeclaration))
+	{
```
Brackets can be removed

leonardoalt commented on test/libsolidity/FunctionCallGraphTest.cpp:
```diff
@@ -0,0 +1,271 @@
+	for (auto const* funcDef: *set)
+		_functions.at(funcDef->name()) = true;
+
+	for(auto pair: _functions)
```
```suggestion
	for (auto pair: _functions)
```

leonardoalt commented on test/libsolidity/FunctionCallGraphTest.cpp:
```diff
@@ -0,0 +1,271 @@
+// Checks that the given _funcDef references all functions in _functions
+void checkRefs(Declaration const* _def, FunctionCallGraphBuilder const& _builder, map<string, bool> _functions)
+{
+	auto* set = _builder.getGraph(_def);
```
```suggestion
	auto const* set = _builder.getGraph(_def);
```
?

leonardoalt commented on test/libsolidity/FunctionCallGraphTest.cpp:
```diff
@@ -0,0 +1,271 @@
+		BOOST_REQUIRE_MESSAGE(pair.second, "Function " + pair.first + " missing.");
+}
+
+Declaration const* findDef(ContractDefinition const* _contract, string _name)
```
```suggestion
Declaration const* findDef(ContractDefinition const& _contract, string const& _name)
```

leonardoalt reviewed (commented):
Some tiny comments

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `7bcfe20c53f55fc014987fb2d2f0dc6cd3647a77`:
```
Coding style error:
 libsolidity/analysis/ImmutableValidator.cpp:108: if( !m_currentConstructor)

```
Please check that your changes are working as intended.

Marenz commented:
Updated this a bit. No proper review required at this point, this is just the current status. Todo:

* [x] Integrate class in the rest of the code
* [ ] Add tests
* [ ] Make common function for virtual function resolution
* [ ] cleanup

chriseth commented on libsolidity/analysis/TypeChecker.h:
```diff
@@ -158,7 +158,7 @@ class TypeChecker: private ASTConstVisitor
 	bool expectType(Expression const& _expression, Type const& _expectedType);
 	/// Runs type checks on @a _expression to infer its type and then checks that it is an LValue.
-	void requireLValue(Expression const& _expression);
+	void requireLValue(Expression const& _expression, bool _valueIsRead = false);
```
Please don't add a default - we have so few call places.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1488,7 +1488,7 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
 	bool const modifying = (op == Token::Inc || op == Token::Dec || op == Token::Delete);
 	if (modifying)
-		requireLValue(_operation.subExpression());
+		requireLValue(_operation.subExpression(), op == Token::Inc || op == Token::Dec);
```
`delete` does not actually read, but it should also not serve as a regular assignment for immutables. Maybe we change the flag instead to denote "ordinary assignment or not"?

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -0,0 +1,88 @@
+namespace solidity::frontend
+{
+
+// Traverses the constructors and state variable assignments to builds a
```
Wrong comment type.

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -0,0 +1,88 @@
+
+		bool operator()(Declaration const* _lhs, Declaration const* _rhs) const
+		{
+			return _lhs->id() < _rhs->id();
```
Can we please move that to ASTNode as the default comparison operator?

chriseth replied:
Ah, the problem is that it is a pointer, right? Then at least add a struct `CompareByID` to ASTNode or something.

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -0,0 +1,88 @@
+		m_currentContract(_contractDefinition),
+		m_errorReporter(_errorReporter)
+	{
+		solAssert(_contractDefinition, "contract must be non-null.");
```
Please pass by reference instead.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,225 @@
+				);
+}
+
+CallableDeclaration const* ImmutableValidator::findFinalOverride(CallableDeclaration const* _callable)
```
Can you extract that to a common place and also use it in CompilerContext?

chriseth replied:
Can now use the other implementation.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,237 @@
+void ImmutableValidator::analyze()
+{
+	for (VariableDeclaration const* stateVar: m_currentContract->stateVariablesIncludingInherited())
+		if (stateVar->immutable() && stateVar->value())
```
We also have to visit all non-constant state variables. They value is evaluated at compile-time and can call arbitrary functions.

Marenz commented:
pushed a small update:
* added tests
* fixed comments
* fixed missing null checks at various places
* integrated into the rest of the codebase

still missing:
* common virtual resolution
* resolving of things like `B.myFunction`

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,237 @@
+bool ImmutableValidator::visit(IfStatement const& _ifStatement)
+{
+	bool prevInBranch = m_inBranch;
+	m_inBranch = true;
```
Can be moved below the condition

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,237 @@
+bool ImmutableValidator::visit(WhileStatement const& _whileStatement)
+{
+	bool prevInLoop = m_inLoop;
+	m_inLoop = true;
```
This has to stay here

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,237 @@
+	else
+		m_errorReporter.typeError(
+			_identifier.location(),
+			"Immutable variables cannot be read in the constructor or any function or modifier called by the it.");
```
Grammar

chriseth commented on Changelog.md:
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 Compiler Features:
  * Metadata: Added support for IPFS hashes of large files that need to be split in multiple chunks.
+ * Immutable State Vars: Proper validation of initialization and reading during construction time.
```
Immutable state variables should have a single changelog entry in the "Language features" section.

chriseth commented on test/libsolidity/syntaxTests/immutable/writing_after_initialization_modifier.sol:
```diff
@@ -0,0 +1,12 @@
+    }
+}
+// ----
+// TypeError: (111-112): Immutable variables can only be initialized directly in the constructor.
```
Did we already decide on whether initializing at point of declaration is possible or not?

chriseth replied:
Ah sorry, confused the error.

Marenz replied:
ah, I actually thought it is okay (in which case the error message is wrong and needs to be changed)

chriseth commented on test/libsolidity/syntaxTests/immutable/getter.sol:
```diff
@@ -1,5 +1,5 @@
 contract C {
-    uint immutable public x;
+    uint immutable public x = 1;
```
Do we also have semantic tests for that? I think that still needs to be implemented.

chriseth replied:
Can you create a new pull request just with a test that calls a getter for an immutable variable?

Marenz replied:
Note that this test result is "Unimplemented" so this is indeed still missing. Do you want the semantic test PR anyway?

chriseth replied:
yes, as a reminder that it still needs to be implemenetd.

chriseth replied:
Can you move this test to a different PR and change the expectations so that it fails?

chriseth commented:
Please add a test that uses `super.f`(should visit exactly the super function), and another one that uses `C.f.selector` (should not visit C.f, even if C is a base contract) and one (maybe it's already there) that uses `C.f` where `C` is a base contract (and thus results in an internal call).

chriseth commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -208,6 +208,8 @@ struct ExpressionAnnotation: ASTAnnotation
 	/// Whether the expression is used in a context where the LValue is actually required.
 	bool lValueRequested = false;
+	/// Whether the expression is an ordinary assignment (e.g. false for +=, ++, --, .., but true for +=, =, -=, ..)
+	bool ordinaryLAssignment = true;
```
```suggestion
	bool lValueOfOrdinaryAssignment = false;
```

the expression is NOT an assigment, it is an identifier on the left side of an assignment.

chriseth commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -208,6 +208,8 @@ struct ExpressionAnnotation: ASTAnnotation
 	bool isLValue = false;
 	/// Whether the expression is used in a context where the LValue is actually required.
 	bool lValueRequested = false;
+	/// Whether the expression is an ordinary assignment (e.g. false for +=, ++, --, .., but true for +=, =, -=, ..)
```
Should also be false for `+=` and `delete`. Only true for `=`

chriseth commented on libsolidity/ast/AST.h:
```diff
@@ -62,6 +62,24 @@ class ASTConstVisitor;
+		{
+			return _lhs->id() < _rhs->id();
+		}
+		bool operator()(ASTNode const* _lhs, int64_t _rhs) const
```
Do we need those?

Marenz replied:
They allow searching in the set using only the id

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1345,7 +1345,7 @@ void TypeChecker::checkExpressionAssignment(Type const& _type, Expression const&
 bool TypeChecker::visit(Assignment const& _assignment)
 {
-	requireLValue(_assignment.leftHandSide());
+	requireLValue(_assignment.leftHandSide(), false);
```
```suggestion
	requireLValue(_assignment.leftHandSide(), _assignment.assignmentOperator() != Token::Assign);
```

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1403,7 +1403,7 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
 			if (component)
 			{
-				requireLValue(*component);
+				requireLValue(*component, false);
```
```suggestion
				requireLValue(*component, _tuple.annotation().lValueOfOrdinaryAssignment);
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+		m_inConstructionContext = false;
+
+		for (auto funcDef: contract->definedFunctions())
+			if (m_visitedCallables.find(funcDef) == m_visitedCallables.end())
```
`contains`

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+			if (auto args = inheritSpec->arguments())
+				ASTNode::listAccept(*args, *this);
+
+		m_inConstructionContext = false;
```
Is it essential that you first visit all functions of one contract instead of visiting all constructors first and then everything else?

chriseth replied:
I think it would be cleaner to first visit everything that can be called from construction context and mark it as already visited and only then visit everything else.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+
+	if (auto funcType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
+		if (
+			(funcType->kind() == FunctionType::Kind::Internal ||
```
coding style

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+		if (
+			(funcType->kind() == FunctionType::Kind::Internal ||
+			funcType->kind() == FunctionType::Kind::Declaration) &&
+			funcType->hasDeclaration()
```
Is it possible that this is false?

Marenz replied:
Yeah, it was false for https://github.com/ethereum/solidity/blob/develop/test/libsolidity/syntaxTests/memberLookup/internal_function_type.sol

chriseth replied:
Ah right!

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+			funcType->kind() == FunctionType::Kind::Declaration) &&
+			funcType->hasDeclaration()
+		)
+			if (m_visitedCallables.insert(dynamic_cast<CallableDeclaration const*>(&funcType->declaration())).second)
```
Can you create a utility function for that? (handling "already visited and calling accept if not" and "setting visited"?)

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+		if (!m_currentConstructor)
+			m_errorReporter.typeError(
+				_identifier.location(),
+				"Immutable variables can only be initialized directly in the constructor.");
```
coding style

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+	else if (m_inConstructionContext)
+		m_errorReporter.typeError(
+			_identifier.location(),
+			"Immutable variables cannot be read in the constructor or any function or modifier called by it.");
```
```suggestion
			"Immutable variables cannot be read in the constructor or in any function or modifier called from it.");
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,252 @@
+
+		for (auto funcDef: contract->definedFunctions())
+			if (m_visitedCallables.find(funcDef) == m_visitedCallables.end())
+				funcDef->accept(*this);
```
Shouldn't this also set it to be already visited?

chriseth commented:
Do we have tests where an immutables are initialized in the following way:
```
constructor() A(x = 2) {}
```
(where `A` is a base contract)

```
constructor() mod(x = 2) {}
```
(where `mod` is a modifier)

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `862d00b868ee0c930df05f5fceecdf1741963c3c`:
```
Coding style error:
 libsolidity/analysis/ImmutableValidator.cpp:18:#include "liblangutil/SourceLocation.h"

```
Please check that your changes are working as intended.

Marenz commented:
Pushed a big update that should address all comments. All in an extra commit that will be amended once approved.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1403,7 +1406,10 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
-				requireLValue(*component);
+				requireLValue(
+					*component,
+					_tuple.annotation().lValueOfOrdinaryAssignment
```
Is there a test for this? `(x, y) = f()` where `x` and `y` are immutables?

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -0,0 +1,80 @@
+namespace solidity::frontend
+{
+
+/* Validates access and initialization of immutable variables:
```
```suggestion
/**
 * Validates access and initialization of immutable variables:
```

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -0,0 +1,80 @@
+*/
+class ImmutableValidator: private ASTConstVisitor
+{
+	using CallSet = std::set<CallableDeclaration const*, ASTNode::CompareByID>;
```
I would prefer `CallableDeclarationSet`, but please at least `CallableSet`.

chriseth commented:
Please rebase on top of develop and use the function introduced in https://github.com/ethereum/solidity/pull/8533

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+	m_readingOfImmutableAllowed = false;
+	m_initializationOfImmutableAllowed = true;
+
+	for (VariableDeclaration const* stateVar: m_currentContract.stateVariablesIncludingInherited())
```
This excludes private state variables. Please add a test and change.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+			solAssert(m_initializedStateVariables.emplace(stateVar).second, "");
+		}
+
+	auto linearizeContracts = m_currentContract.annotation().linearizedBaseContracts  | boost::adaptors::reversed;
```
```suggestion
	auto linearizedContracts = m_currentContract.annotation().linearizedBaseContracts | boost::adaptors::reversed;
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+
+	auto linearizeContracts = m_currentContract.annotation().linearizedBaseContracts  | boost::adaptors::reversed;
+
+	for (auto const* contract: linearizeContracts)
```
```suggestion
	for (ContractDefinition const* contract: linearizeContracts)
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+
+	m_initializationOfImmutableAllowed = false;
+
+	for (auto const* contract: linearizeContracts)
```
```suggestion
	for (ContractDefinition const* contract: linearizeContracts)
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+		if (funcDef->isConstructor())
+			m_currentConstructor = funcDef;
+
+		// Disallow init. in the args of mods/base'ctor calls
```
Please use fewer abbreviations, this is very hard to read.

chriseth commented on libsolidity/analysis/ImmutableValidator.h:
```diff
@@ -0,0 +1,80 @@
+	FunctionDefinition const* m_currentConstructor = nullptr;
+	bool m_inLoop = false;
+	bool m_inBranch = false;
+	bool m_readingOfImmutableAllowed = false;
```
Are they not always the opposite of each other? Also "initializationAllowed" does not really mean that is allowed, it just means it might be allowed, right?

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+		if (contract->constructor())
+			visitCallable(*contract->constructor());
+
+	m_initializationOfImmutableAllowed = false;
```
Is this really needed? We are not inside a constructor anyway.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+				);
+}
+
+void ImmutableValidator::visitCallable(Declaration const& _declaration)
```
```suggestion
void ImmutableValidator::visitCallableIfNotAlreadyVisited(Declaration const& _declaration)
```
?

chriseth replied:
`visitCallableIfUnknown`?

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+
+bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+{
+	if (
```
So is this actually needed in the end?

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+		dynamic_cast<FunctionType const*>(_memberAccess.expression().annotation().type) &&
+		dynamic_cast<FixedBytesType const*>(_memberAccess.annotation().type)
+	)
+		return false;
```
Even if this stays, we should visit further inwards, we could have something like `f().selector`, where `f` returns a function pointer.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+	return analyseCallable(_modifierDefinition);
+}
+
+bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
```
I'm wondering if it would be better to not implement any visitor for the expression nodes, but instead override the `visitNode(ASTNode const& _node)` function and check if `_node` is an expression.

chriseth replied:
(although implementing the visitor for Identifier makes sense)

chriseth replied:
Ah sorry, `referencedDeclaration` only exists for `MemberAccess` and `Identifier`, so it's fine as it is!

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+	return false;
+}
+
+bool ImmutableValidator::visit(Return const& _return)
```
Would it be simpler to make this an `endVisit` and call `checkAll...` only in case `m_currentContructor` is not null?

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+
+		// Disallow init. in the args of mods/base'ctor calls
+		bool previousInitAllowed = m_initializationOfImmutableAllowed;
+		m_initializationOfImmutableAllowed = false;
```
Wouldn't setting the current constructor only after visiting the modifiers also be a solution?

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+	if (FunctionDefinition const* funcDef = dynamic_cast<decltype(funcDef)>(&_callableDeclaration))
+	{
+		if (funcDef->isConstructor())
+			m_currentConstructor = funcDef;
```
Shouldn't you reset the current constructor in the else case?

chriseth replied:
Ah sorry, it is set to nullptr in line 154

Marenz replied:
It is already set to null at the beginning of this function

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+	return false;
+}
+
+void ImmutableValidator::analyseVariableDeclaration(VariableDeclaration const& _variableDeclaration, Expression const& _expression)
```
```suggestion
void ImmutableValidator::analyseVariableReference(VariableDeclaration const& _referencedVariable, Expression const& _expression)
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+		if (!m_currentConstructor)
+			m_errorReporter.typeError(
+				_expression.location(),
+				"Immutable variables can only be initialized inline or directly in the constructor."
```
```suggestion
				"Immutable variables can only be initialized inline or assigned directly in the constructor."
```

chriseth replied:
In case this is a regular assignment, people might not see the connection to the place of assignment. ("Yes, I know they can only be initialized in the constructor, but I'm not initializing, I'm assigning!")

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+
+void ImmutableValidator::checkAllVariablesInitialized(langutil::SourceLocation const& _location)
+{
+	for (VariableDeclaration const* varDecl: m_currentContract.stateVariablesIncludingInherited())
```
Same here, this excludes private ones.

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+				m_errorReporter.typeError(
+					_location,
+					langutil::SecondarySourceLocation().append("Not initialized: ", varDecl->location()),
+					"Construction controlflow ends without initializing all immutable state variables."
```
```suggestion
					"Construction control flow ends without initializing all immutable state variables."
```

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,271 @@
+			visitCallable(*modDef);
+	}
+
+	checkAllVariablesInitialized(m_currentContract.location());
```
I know this is also about initializing inline, but shouldn't we reference the constructor location if there is a constructor? The text talks about "constructor control flow".

Marenz replied:
I changed it to `construction work flow` which is a bit more generic. To me referencing the c'tor here could be confusing

chriseth commented:
Even though I had many comments, this looks really nice!

chriseth commented on libsolidity/analysis/ImmutableValidator.cpp:
```diff
@@ -0,0 +1,214 @@
+		analyseVariableReference(*varDecl, _memberAccess);
+	else if (auto funcType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
+		if (funcType->kind() == FunctionType::Kind::Internal && funcType->hasDeclaration())
+			visitCallableIfNew(funcType->declaration());
```
Shouldn't this use `resolveVirtual`?

chriseth replied:
Ah no, it's not done here.

chriseth commented:
Looks good apart from the failing tests!

Marenz committed:
Always return `this` for `EventDefinition::resolveVirtual`

Instead of asserting.

chriseth commented:
Build error on windows:
```
C:\projects\solidity\libsolidity\analysis\ImmutableValidator.cpp(194): error C2653: 'langutil': is not a class or namespace name [C:\projects\solidity\build\libsolidity\solidity.vcxproj]
C:\projects\solidity\libsolidity\analysis\ImmutableValidator.cpp(194): error C2065: 'SourceLocation': undeclared identifier [C:\projects\solidity\build\libsolidity\solidity.vcxproj]
C:\projects\solidity\libsolidity\analysis\ImmutableValidator.cpp(194): error C2059: syntax error: 'const' [C:\projects\solidity\build\libsolidity\solidity.vcxproj]
C:\projects\solidity\libsolidity\analysis\ImmutableValidator.cpp(195): error C2143: syntax error: missing ';' before '{' [C:\projects\solidity\build\libsolidity\solidity.vcxproj]
C:\projects\solidity\libsolidity\analysis\ImmutableValidator.cpp(195): error C2447: '{': missing function header (old-style formal list?) [C:\projects\solidity\build\libsolidity\solidity.vcxproj]
```

chriseth commented on test/libsolidity/syntaxTests/immutable/getter.sol:
```diff
@@ -1,5 +1,5 @@
+    uint immutable public x = 1;
 }
 // ----
 // UnimplementedFeatureError: NONE
```
Please move this test out of this PR.

Marenz replied:
The test already exists in the repo, I am just fixing it so it doesn't fail. Should I remove it completely?

chriseth replied:
Yes, please move it to its own PR. We should not have tests that are expected to be failing.
